client/webserver/api.go

```diff
@@ -116,6 +116,29 @@ func (s *WebServer) apiOpenWallet(w http.ResponseWriter, r *http.Request) {
 	writeJSON(w, simpleAck(), s.indent)
 }
 
+// apiNewDepositAddress gets a new deposit address from a wallet.
+func (s *WebServer) apiNewDepositAddress(w http.ResponseWriter, r *http.Request) {
+	form := &struct {
+		AssetID uint32 `json:"assetID"`
+	}{}
+	if !readPost(w, r, form) {
+		return
+	}
```

Comment:
Interesting consequence of json unmarshalling here:

```go
func Test_readPost(t *testing.T) {
	form := &struct {
		AssetID uint32 `json:"assetID"`
	}{}
	err := json.Unmarshal([]byte(`{}`), form)
	if err != nil {
		t.Fatal(err)
	}
	t.Log("assetID:", form.AssetID) // assetID is 0 (bitcoin)
}
```

Even though the body JSON does not have a `"assetID"` field, this would get a new deposit address for asset 0 (btc).

Maybe this is OK.  If this could backfire, we might want to have the form struct have a `*uint32` so we can check if that's nil.

e.g.

```go
func TestGetOrderIDCtx(t *testing.T) {
	oid := encode.RandomBytes(32)
	hexOID := hex.EncodeToString(oid)

	r := (&http.Request{}).WithContext(context.WithValue(context.Background(), ctxOID, hexOID))

	bytesOut, err := getOrderIDCtx(r)
	if err != nil {
		t.Fatalf("getOrderIDCtx error: %v", err)
	}
	if len(bytesOut) == 0 {
		t.Fatalf("empty oid")
	}
	if !bytes.Equal(oid, bytesOut) {
		t.Fatalf("wrong bytes. wanted %x, got %s", oid, bytesOut)
	}

	// Test some negative paths
	for name, v := range map[string]interface{}{
		"nil":          nil,
		"int":          5,
		"wrong length": "abc",
		"not hex":      "zyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxwzyxw",
	} {
		r := (&http.Request{}).WithContext(context.WithValue(context.Background(), ctxOID, v))
		_, err := getOrderIDCtx(r)
		if err == nil {
			t.Fatalf("no error for %v", name)
		}
	}
}

func Test_readPost(t *testing.T) {
	form := &struct {
		AssetID *uint32 `json:"assetID"`
	}{}
	err := json.Unmarshal([]byte(`{}`), form)
	if err != nil {
		t.Fatal(err)
	}
	if form.AssetID != nil {
		t.Fatal("unexpectedly set")
	}

	form = &struct {
		AssetID *uint32 `json:"assetID"`
	}{}
	err = json.Unmarshal([]byte(`{"assetID":0}`), form)
	if err != nil {
		t.Fatal(err)
	}
	if form.AssetID == nil {
		t.Fatal("should have been set")
	}
	t.Log("assetID:", *form.AssetID)
}
```

Maybe it's not worth complicating.  Something to think about.

Reply:
Pasted too much there, sorry.  Cleaned up the last comment.